From: Antoine Musso <hashar@free.fr>
Date: Mon, 15 Jan 2018 16:00:14 +0100
Subject: Hook to run lintian

We better use the lintian version of the target distribution instead of
the host lintian version.

Change-Id: I01137a1f621061f5063c8624b9e6eaac44405d8d
---
 pbuilder-hookdir/B90lintian       | 40 +++++++++++++++++++++++++++++++++++++++
 scripts/build-and-provide-package |  6 ++++++
 2 files changed, 46 insertions(+)
 create mode 100755 pbuilder-hookdir/B90lintian

diff --git a/pbuilder-hookdir/B90lintian b/pbuilder-hookdir/B90lintian
new file mode 100755
index 0000000..5d52f1c
--- /dev/null
+++ b/pbuilder-hookdir/B90lintian
@@ -0,0 +1,40 @@
+#!/bin/bash
+#
+# Hook to run lintian inside the cowbuilder environment.
+#
+
+BUILDDIR="${BUILDDIR:-/tmp/buildd}"
+
+if [ "${LINTIAN:-}" != "true" ] ; then
+  echo "Skipping lintian (LINTIAN is not 'true')"
+  exit 0
+fi
+
+if [ -n "${LINTIAN_OPTIONS:-}" ] ; then
+  echo "*** Using provided LINTIAN_OPTIONS $LINTIAN_OPTIONS ***"
+fi
+
+set -ex -o pipefail
+apt-get -y "${APTGETOPT[@]}" install lintian
+lintian --version
+
+PACKAGE_NAME=$(basename "$(realpath "$BUILDDIR"/*/debian/..)")
+echo "Found package name: $PACKAGE_NAME"
+
+# Requires a .pbuilderrc with:
+#  export ADDITIONAL_BUILDRESULTS=(../*.lintian.txt)
+lintian_out=${BUILDDIR}/${PACKAGE_NAME}.lintian.txt
+su -c "lintian ${LINTIAN_OPTIONS:-} \"${BUILDDIR}\"/*.changes 2>&1 | tee \"${lintian_out}\"" - pbuilder \
+  || EXIT=$?
+# We strip ANSI sequences in case lintian got --color forced
+sed -i 's%\x1B\[[0-9;]*\?[mGKH]%%g' "${lintian_out}"
+
+case ${EXIT:-0} in
+  2)
+    echo "lintian run-time error."
+    exit 2
+    ;;
+  1|*)
+    exit 0
+    ;;
+esac
diff --git a/scripts/build-and-provide-package b/scripts/build-and-provide-package
index 4dc7d25..2783c30 100755
--- a/scripts/build-and-provide-package
+++ b/scripts/build-and-provide-package
@@ -138,6 +138,12 @@ checks_and_defaults() {
     FREIGHT_VARCACHE=${FREIGHT_BASE}/${FREIGHT_REPOSITORY}
     FREIGHT_CONF=${FREIGHT_BASE}/${FREIGHT_REPOSITORY}.conf
   fi
+
+  if [ "${LINTIAN:-}" = 'true' ] ; then
+    # Will run lintian and we need pbuilder to copy the result file to
+    # $BUILDRESULT.
+    ADDITIONAL_BUILDRESULTS+=('../*lintian.txt')
+  fi
 }
 
 clean_workspace() {
